File: examples/python/task-management/api-service/src/external/email_client.py
"""HTTP client for email service communication."""
import httpx
import os
from typing import Dict, Any

class EmailClient:
    """Client for email service API."""

    def __init__(self):
        self.base_url = os.getenv("EMAIL_SERVICE_URL", "http://localhost:3001")
        self.client = httpx.Client(base_url=self.base_url, timeout=10.0)

    def send_task_completed(self, task_id: int, user_id: int) -> Dict[str, Any]:
        """Send task completion email."""
        response = self.client.post(
            "/api/emails/send",
            json={
                "type": "task_completed",
                "task_id": task_id,
                "user_id": user_id
            }
        )
        response.raise_for_status()
        return response.json()

    def send_welcome_email(self, user_id: int, email: str) -> Dict[str, Any]:
        """Send welcome email to new user."""
        response = self.client.post(
            "/api/emails/send",
            json={
                "type": "welcome",
                "user_id": user_id,
                "to": email
            }
        )
        response.raise_for_status()
        return response.json()

File: examples/csharp/task-management/api-service/src/External/EmailClient.cs
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace TaskManagementApi.External;

/// <summary>
/// Client for sending emails via external service.
/// </summary>
public class EmailClient
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<EmailClient> _logger;
    private readonly string _apiKey;

    public EmailClient(HttpClient httpClient, ILogger<EmailClient> logger, IConfiguration configuration)
    {
        _httpClient = httpClient;
        _logger = logger;
        _apiKey = configuration["EmailService:ApiKey"] ?? string.Empty;
    }

    public async Task<bool> SendTaskNotificationAsync(string email, string taskTitle)
    {
        _logger.LogInformation("Sending task notification to {Email}", email);

        var payload = new
        {
            to = email,
            subject = "Task Update",
            body = $"Task '{taskTitle}' has been updated"
        };

        var content = new StringContent(
            JsonSerializer.Serialize(payload),
            Encoding.UTF8,
            "application/json");

        content.Headers.Add("X-API-Key", _apiKey);

        try
        {
            var response = await _httpClient.PostAsync("/api/send", content);
            return response.IsSuccessStatusCode;
        }
        catch (HttpRequestException ex)
        {
            _logger.LogError(ex, "Failed to send email notification");
            return false;
        }
    }

    public async Task<bool> SendWelcomeEmailAsync(string email, string username)
    {
        _logger.LogInformation("Sending welcome email to {Email}", email);

        var payload = new
        {
            to = email,
            subject = "Welcome to Task Manager",
            body = $"Hello {username}, welcome to our task management system!"
        };

        var content = new StringContent(
            JsonSerializer.Serialize(payload),
            Encoding.UTF8,
            "application/json");

        content.Headers.Add("X-API-Key", _apiKey);

        try
        {
            var response = await _httpClient.PostAsync("/api/send", content);
            return response.IsSuccessStatusCode;
        }
        catch (HttpRequestException ex)
        {
            _logger.LogError(ex, "Failed to send welcome email");
            return false;
        }
    }
}
